### PR TITLE
Fixes #10111 - Use a dummy primary interface for unmanaged hosts

### PR DESCRIPTION
**Pull request by dLobatog**
Unmanaged hosts require a host interface due to delegation of networking
attributes, and some others like 'name'. Since unmanaged hosts do not
necessarily have an interface associated (unless created through puppet
facts importer), we provide a dummy interface to prevent errors.

Ideally this should go away as soon as unmanaged hosts are a different
class or Host::Base, moving validations and other particularities
to Host::Managed.

### CODE REVIEW

File: app/controllers/hosts_controller.rb [hunk 1]
```diff
@@ -96,13 +97,6 @@ def edit
   def update
     forward_url_options
     Taxonomy.no_taxonomy_scope do
-      # remove from hash :root_pass and bmc :password if blank?
```
Comment (dLobatog) [inline app/controllers/hosts_controller.rb:99]: This part is not strictly related to the pull request. Just a small cleanup as that part of `update` is a bit hard to parse at first glance.

File: app/models/host/base.rb [hunk 2]
```diff
@@ -246,6 +246,10 @@ def overwrite=(value)
     end
 
     def primary_interface
+      # Returns a dummy object if host is unmanaged and without interfaces.
+      # Unmanaged hosts still have the interface requirement due to delegations and certain validations.
+      # A refactoring that allows unmanaged hosts to exist without interfaces would be welcome.
+      return Nic::Base.new(:host => self, :primary => true) unless managed && interfaces.present?
```
Comment (dLobatog) [inline app/models/host/base.rb]: We pass `:host => self` and  `:primary => true` so that [NameSynchronizer](https://github.com/theforeman/foreman/blob/develop/app/services/name_synchronizer.rb) sets the name of the host properly. 

_....rant....._

Yes.

I also think it couldn't be dirtier but _1.8_ is near and we won't have another RC for testing.

File: app/models/host/base.rb [hunk 3]
```diff
@@ -246,6 +246,10 @@ def overwrite=(value)
     end
 
     def primary_interface
+      # Returns a dummy object if host is unmanaged and without interfaces.
+      # Unmanaged hosts still have the interface requirement due to delegations and certain validations.
+      # A refactoring that allows unmanaged hosts to exist without interfaces would be welcome.
+      return Nic::Base.create(:host => self, :primary => true) unless new_record? || managed || interfaces.present?
```
Comment (dLobatog) [inline app/models/host/base.rb]: We pass :host => self and :primary => true so that NameSynchronizer sets the name of the host properly.

....rant.....

Yes.

I also think it couldn't be dirtier but 1.8 is near and we won't have another RC for testing.

File: app/models/host/base.rb [hunk 4]
```diff
@@ -246,7 +246,11 @@ def overwrite=(value)
     end
 
     def primary_interface
-      get_interface_by_flag(:primary)
+      # Returns a dummy object if host is unmanaged and without interfaces.
+      # Unmanaged hosts still have the interface requirement due to delegations and certain validations.
+      # A refactoring that allows unmanaged hosts to exist without interfaces would be welcome.
+      return get_interface_by_flag(:primary) if managed || interfaces.detect(&:primary)
+      self.interfaces.build(:primary => true, :provision => true, :managed => false, :type => 'Nic::Managed')
```
Comment (dLobatog) [inline app/models/host/base.rb]: This is pretty much equivalent to the `build_required_interfaces` method called in the constructor. 

....rant.....

Yes.

I also think it couldn't be dirtier but 1.8 is near and we won't have another RC for testing.

File: app/models/host/base.rb [hunk 5]
```diff
@@ -246,7 +246,11 @@ def overwrite=(value)
     end
 
     def primary_interface
-      get_interface_by_flag(:primary)
+      # Returns a dummy object if host is unmanaged and without interfaces.
+      # Unmanaged hosts still have the interface requirement due to delegations and certain validations.
+      # A refactoring that allows unmanaged hosts to exist without interfaces would be welcome.
+      return get_interface_by_flag(:primary) if managed || interfaces.detect(&:primary)
```
Comment (ohadlevy) [inline app/models/host/base.rb]: nitpick: if managed?

**Comment by domcleal** [comment]
@ares could you look at this too please?  I'm undecided if the dirtiness for now is unavoidable or not...

**Comment by dLobatog** [comment]
@tstrachota a review would help too since @ares is not available this week

**Comment by tstrachota** [comment]
@eLobato I'm checking it now.

File: app/models/host/base.rb [hunk 6]
```diff
@@ -246,7 +246,11 @@ def overwrite=(value)
     end
 
     def primary_interface
-      get_interface_by_flag(:primary)
+      # Returns a dummy object if host is unmanaged and without interfaces.
+      # Unmanaged hosts still have the interface requirement due to delegations and certain validations.
+      # A refactoring that allows unmanaged hosts to exist without interfaces would be welcome.
+      return get_interface_by_flag(:primary) if managed? || interfaces.detect(&:primary)
+      self.interfaces.build(:primary => true, :provision => true, :managed => false, :type => 'Nic::Managed')
```
Comment (tstrachota) [inline app/models/host/base.rb]: There's a method `build_required_interfaces` https://github.com/theforeman/foreman/blob/develop/app/models/host/base.rb#L325 that does the same as the line above plus it checks existence of provision interface. Could you use it instead?

File: app/models/host/base.rb (hunk 4 above)
Comment (tstrachota) [inline app/models/host/base.rb]: Ah, I see your previous comment now. What was the reason for not using `build_required_interfaces` here too?

**Comment by tstrachota** [comment]
One question in the code. It worked ok in my tests. :+1: for this workaround.

**Comment by dLobatog** [comment]
@tstrachota @domcleal Updated!

File: app/models/host/base.rb [hunk 7]
```diff
@@ -323,8 +327,8 @@ def tax_organization
     end
 
     def build_required_interfaces(attrs = {})
-      self.interfaces.build(attrs.merge(:primary => true, :type => 'Nic::Managed')) if self.primary_interface.nil?
-      self.primary_interface.provision = true if self.provision_interface.nil?
+      self.interfaces.build(attrs.merge(:primary => true, :type => 'Nic::Managed')) if get_interface_by_flag(:primary).nil?
```
Comment (dLobatog) [inline app/models/host/base.rb:330]: Calling `.primary_interface` here would result on an infinite loop here so I changed it.

**Comment by domcleal** [comment]
Thanks @eLobato, merged as a5dc3e2.  If you can file a Redmine ticket about how you think it could be refactored, that'd be useful too.

**Comment by ares** [comment]
Thanks for fixing. I think we should later allow full interface management for unmanaged hosts, unmanaged host does not mean it does not have networking attributes.

**Comment by domcleal** [comment]
Yeah, I kind of agree.  For one, I was thinking about the lack of storage for the IP address that's now missing in 1.8.  e.g. I used to rely on the IP field on unmanaged hosts, as I'd load up Foreman, view a host and then SSH to the IP it reported.

**Comment by ares** [comment]
Reported as http://projects.theforeman.org/issues/10260
